### ISSUE TITLE
Suggestions

### ISSUE THREAD

**Issue by garethhu** [Khum1/data-collection-pipeline#3]
### book.py

Using the Single Responsibility Principal (SRP), you should separate ('abstract' is the technical term) the object class, from the method of reading data. We always want to abstract our 'domain logic' (what our application does) from its 'integrations' (what our application uses to do it), this is for a number of reasons:

* It allows us to better reason about domain logic when we don't need to think about the interactions with other systems
* It allows us to test the domain logic more easily using unit tests
* It allows us to test the integrations more easily with integration tests
* We can change the integrations more easily without changing our domain logic, for example, if a library or API we are using changes, or a better one becomes available
* We can add new integrations, for example, we could add multiple website sources for our book data, without changing our domain logic
* It is easier to work in parallel with other people on the same project when the code is loosely coupled
* It is easier to troubleshoot problems when we can isolate them to either the integration or domain layer
* We can protect our domain layer from dodgy IO (Input/Output) so that we only have to deal with clean, valid data in our domain logic

To do this you should seek to wrap all uses of 'driver' and related operations to a separate class, the 'Book' class should not care about how the data is obtained, which means it shouldn't have any reference to html or the structure of the document. Additionally, it should not care about how the data is stored locally, so the image storage should also be handled in a separate class.

Additionally, after you have separated out the integration layer from your domain logic you should improve your exception handling, you should never use just pass in an except block. Also, you should be specific in the type of exception you are catching, this is so that we use the correct strategy depending on the specific error that occurred and we don't lose information about the error, or the relevant context in which it occurred. When handling exceptions in an except block, in general: 

* If the behavior is recoverable, it should be recovered, you will also want to [log here at the warning level](https://realpython.com/python-logging/)
* If it is ignorable, then it should be logged at the warning level
* If it isn't recoverable and is due to user error (this should never be critical) then you should throw a case-specific subtype of the RuntimeError exception, this should be then handled by some top-level handler that catches these exceptions and returns a coherent error message to the user
* If it isn't recoverable, isn't due to user error, and is non critical  then you should be logging at the error level
* If it is a critical failure then you should throw a case-specific subtype of the RuntimeError exception that should at some point at the top level crash the program and log at the critical log level, we would throw upward so that parts of the application can close any operating system resources still in use, we would rather a crash than invalid behavior (it is very rare that you need to do this, most of these cases are already baked into python)
* If it cannot be handled at the current level but may be able to be handled at a higher level, then we should create a new case-specific Error type, throw a new exception of this type and pass in the  error that occurred as a 'cause' for our new domain-specific type

Note that we should try to avoid logging and throwing the exception up, this is so that we don't end up with loads of logs saying the same thing. Also whenever we use or create an exception, we should be as specific as possible in the choice of exception type, and we should pass into it any contextually helpful information that may be helpful during troubleshooting, such as the URL that failed, the name of the book, etc.

### system.py

Name this something better, looks like it is initializing your local file system for use in the app, maybe call it something like FileSystemHandler, the creation of the file system would then be performed in the constructor (init method) with helpful functions for interacting with it, the file system logic could be moved from your other classes into this so you can do file_system_handler.save(some_image) or something similar.

The driver functionality here should be moved into a class that handles the driver interactions, the same one I mentioned creating in book.py.

### scraper.py

As your program entry point, I think there should be a `if __name__ == "__main__":` wrapping the `scrape_website` invocation so that it is only run as part of the program invocation, I am not sure on this point, it's a while since I did python.

The same issue regarding exception handling is here, as mentioned above.

Move the driver stuff out to the driver wrapper I've previously talked about.

Try to reduce the sleep statements if possible, sleeps are generally bad practice. They may be here for a good reason though, such as reducing the request rate to the website so there is no throttling or flagging of programmatic usage, not sure if you have been told to use them.

Try using a [logging package](https://realpython.com/python-logging/) instead of `print` this is good industry practice.

I am not sure what is happening here with URL, it is used in multiple places to initialize the driver, ideally, you should store the URL in a [config file](https://martin-thoma.com/configuration-files-in-python/), though this may be out of scope. Generally speaking, any 'special values' that are not universal constants such as seconds in a minute should be put in config files. Also, the second usage of URL in the for loop is capitalized, we should only capitalize constant values, even though it is an acronym because the value changes we should use the lowercase `url`.

I have other comments and tips, but probably best to leave them for now, since they can be more specific after you have made some of the above changes. If anything is unclear let me know.
